### PR TITLE
Feature/perl install

### PR DESCRIPTION
**Pull request by jemten**

### CODE REVIEW

File: perl_install.pl [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+    q{pma|perl_modules_append:s}    => \@{ $parameter{perl_modules_append} },
+    q{ppd|print_parameters_default} => sub {
+        print_parameters(
+            {
+                parameter_href       => \%parameter,
+                array_parameter_href => \%array_parameter
+            }
+        );
+        exit;
+    },    # Display parameter defaults
```
Comment (henrikstranneheim) [inline perl_install.pl]: move comment to above sub call

File: perl_install.pl [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+    q{q|quiet} => \$parameter{quiet},
+    q{h|help}  => sub {
+        say STDOUT $USAGE;
+        exit;
+    },    #Display help text
```
Comment (henrikstranneheim) [inline perl_install.pl]: move comment to above say STDOUT

File: perl_install.pl [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+    say {$FILEHANDLE} q{## Skipping installation of perl};
+}
+elsif ( $PERL_VERSION ge q{v} . $parameter{perl_version}
+    and not $parameter{perl_force_install} )
+{
+    say STDERR q{## Perl version requirements already met, }
+      . q{current perl version: }
+      . $PERL_VERSION;
+    say STDERR
+      q{## Re-run the script with '-pfi' flag to force perl installation};
```
Comment (henrikstranneheim) [inline perl_install.pl]: Always use long description i.e. 'perl_force_install' when helping user. It easier to understand

File: perl_install.pl [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+            parameter_href => \%parameter,
+            FILEHANDLE     => $FILEHANDLE
+        }
+    );
+}
+say {$FILEHANDLE} qq{\n};
+
+## Write cpanm install instructions
+# Check if cpanm alread in path
+my ($inst)    = ExtUtils::Installed->new();
```
Comment (henrikstranneheim) [inline perl_install.pl]: $inst = $installation??

Comment (henrikstranneheim) [reply perl_install.pl]: Installed? Would be nice with my (@modules) = $installed->modules();

File: perl_install.pl [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+            parameter_href => \%parameter,
+            FILEHANDLE     => $FILEHANDLE
+        }
+    );
+}
+say {$FILEHANDLE} qq{\n};
+
+## Install MIP requred modules via cpanm
+say {$FILEHANDLE} q{## Install modules reuired by MIP via cpanm};
+install_modules(
```
Comment (henrikstranneheim) [inline perl_install.pl]: Perhaps install_cpanm_modules if it is specific to CPANM

File: perl_install.pl [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+    # Use newly installed perl
+    say {$FILEHANDLE} q{export PATH=} . $perl_install_path . q{/:$PATH};
+
+    # Add at start-up
+    say {$FILEHANDLE} q{echo 'eval `perl -I }
+      . catdir( $perl_install_path, qw{ lib perl5 } )
+      . q{ -Mlocal::lib=}
+      . $perl_install_path
+      . q{/`' >> ~/.bash_profile } . qq{\n};
+
```
Comment (henrikstranneheim) [inline perl_install.pl]: Handle Unicode

File: perl_install.pl [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+## Function  : Perl wrapper for writing cpanm recipe to $FILEHANDLE.
+## Returns   : ""
+## Arguments : $parameter_href, $FILEHANDLE
+##           : $parameter_href => Hash with paramters {REF}
+##           : $FILEHANDLE     => Filehandle to write to
+
+    my ($arg_href) = @_;
+
+    $parameter_href = $arg_href->{parameter_href};
```
Comment (henrikstranneheim) [inline perl_install.pl]: It would be nice if you flattened the arguments i.e. $force = $arg_href->{parameter_href}{force}

File: perl_install.pl [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+##           : $FILEHANDLE     => Filehandle to write to
+
+    my ($arg_href) = @_;
+
+    ## Flatten arguments
+    my $url;
+    my $outfile_path;
+
+    ## Load arguments
+    $parameter_href = $arg_href->{parameter_href};
```
Comment (henrikstranneheim) [inline perl_install.pl]: It would be nice if you flattened the arguments i.e. $quiet = $arg_href->{parameter_href}{quiet}

File: perl_install.pl [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+        {
+            parameter_href => $parameter_href,
+            url            => $url,
+            outfile_path   => $outfile_path,
+            FILEHANDLE     => $FILEHANDLE
+        }
+    );
+
+    my $perl_install_path = catdir(
+        $parameter_href->{perl_install_dir},
```
Comment (henrikstranneheim) [inline perl_install.pl]: Import this and flatten:  $perl_install_dir = $arg_href->{parameter_href}{perl_install_dir}

File: perl_install.pl [hunk 10]
```diff
@@ -0,0 +1,595 @@
...
+
+## Function  : Write perl isntallation recipe to FILEHANDLE
+## Returns   : ""
+## Arguments : $parameter_href, $array_parameter_href
+##           : $parameter_href       => Parameters hash {REF}
+##           : $array_parameter_href => Hold the array parameter defaults as {REF}
+
+    my ($arg_href) = @_;
+
+    $parameter_href = $arg_href->{parameter_href};
```
Comment (henrikstranneheim) [inline perl_install.pl]: Import this and flatten e.g.  $perl_install_dir = $arg_href->{parameter_href}{perl_install_dir}

File: lib/MIP/PacketManager/Cpanm_v5_10.pm [hunk 11]
```diff
@@ -0,0 +1,69 @@
+package MIP::PacketManager::Cpanm_v5_10;
```
Comment (henrikstranneheim) [inline lib/MIP/PacketManager/Cpanm_v5_10.pm]: I don't rembemer but should it be PacketManager or Packetmanager or Packet_manager. What does best practise say

File: lib/MIP/PacketManager/Cpanm_v5_10.pm [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    our $VERSION = 1.0.0;
+
+    # Functions and variables which can be optionally exported
+    our @EXPORT_OK = qw{ cpanm_install_module };
+
+}
+
+sub cpanm_install_module {
+
+## cpanm_install
```
Comment (henrikstranneheim) [inline lib/MIP/PacketManager/Cpanm_v5_10.pm]: cpanm_install -> cpanm_install_module

File: perl_install.pl [hunk 13]
```diff
@@ -226,51 +224,55 @@ sub install_perl {
...
         }
     );
 
+    ## Write commands to $FILEHANDLE
+    print {$FILEHANDLE} join( q{ }, @commands ) . q{ };
+
     print {$FILEHANDLE} qq{\n};
 
-    say {$FILEHANDLE} qq{cd } . $parameter_href->{perl_install_dir} . qq{\n};
+    say {$FILEHANDLE} qq{cd } . $perl_install_dir . qq{\n};
```
Comment (henrikstranneheim) [inline perl_install.pl]: why qq{cd } and not q{cd }

File: perl_install.pl [hunk 14]
```diff
@@ -390,9 +400,9 @@ sub install_cpanm {
     return;
 }
 
-sub install_modules {
+sub install_cpanm_modules {
 
-## insatall_modules
+## insatall_cpanm_modules
```
Comment (henrikstranneheim) [inline perl_install.pl]: insatall??

File: lib/MIP/PacketManager/Cpanm_v5_10.pm (hunk 11 above)
Comment (jemten) [reply lib/MIP/PacketManager/Cpanm_v5_10.pm]: Can't find anything but I suggest we change it to Package_manager

**Comment by henrikstranneheim** [comment]
Excellent! Now given time we should update the docs to reflect this change
👍

**Pull request merged**